### PR TITLE
Spring Boot Security: add User entity

### PR DESCRIPTION
**Pull request by dmarmul**
Added a User entity, repository and service, mapper, DTOs, and a registration endpoint. Also added custom annotation for checking that password and repeatPassword field are the same. Used  RegistrationException by the GlobalExceptionHandler.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -111,6 +111,11 @@
             <groupId>com.fasterxml.jackson.core</groupId>
             <artifactId>jackson-databind</artifactId>
         </dependency>
+        <dependency>
+            <groupId>commons-beanutils</groupId>
+            <artifactId>commons-beanutils</artifactId>
```
Comment (Elena-Bruyako) [inline pom.xml]: remove

File: src/main/java/org/example/bookshop/annotation/FieldMatch.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package org.example.bookshop.annotation;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/annotation/FieldMatch.java]:
```suggestion
package org.example.bookshop.validation;
```

File: src/main/java/org/example/bookshop/validation/FieldMatchValidator.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        this.firstFieldName = constraintAnnotation.first();
+        this.secondFieldName = constraintAnnotation.second();
+    }
+
+    @Override
+    public boolean isValid(Object object, ConstraintValidatorContext context) {
+        try {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/validation/FieldMatchValidator.java]: see an example

```
Object field = new BeanWrapperImpl(value).getPropertyValue(this.field);
Object fieldMatch = new BeanWrapperImpl(value).getPropertyValue(this.fieldMatch);
return Objects.equals(field, fieldMatch);
```

File: src/main/java/org/example/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 4]
```diff
@@ -11,28 +11,42 @@
...
-        body.put("errors", errors);
+        body.put(ERRORS, errors);
         return new ResponseEntity<>(body, headers, HttpStatus.BAD_REQUEST);
     }
 
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(
+            RegistrationException ex) {
+        body.put(TIMESTAMP, LocalDateTime.now());
+        body.put(STATUS, HttpStatus.BAD_REQUEST);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/exception/CustomGlobalExceptionHandler.java]:
```suggestion
        body.put(STATUS, HttpStatus.CONFLICT);
```

File: src/main/java/org/example/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 5]
```diff
@@ -11,28 +11,42 @@
...
+        body.put(ERRORS, ex.getMessage());
+        return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/exception/CustomGlobalExceptionHandler.java]:
```suggestion
        return new ResponseEntity<>(body, HttpStatus.CONFLICT);
```

File: src/main/java/org/example/bookshop/repository/UserRepository.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package org.example.bookshop.repository;
+
+import java.util.Optional;
+import org.example.bookshop.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface UserRepository extends JpaRepository<User, Long> {
+    Optional<User> findByEmail(String email);
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/repository/UserRepository.java]:
```suggestion
    boolean existsByEmail(String email);
```

File: src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto save(UserRegistrationRequestDto requestDto) {
+        Optional<User> user = userRepository.findByEmail(requestDto.getEmail());
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java]:
```suggestion
        if (userRepo.existsByEmail(request.email())) {
            throw new RegistrationException(...
        User user = userMapper.toUser(requestDto);
        ....
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/org/example/bookshop/dto/UserResponseDto.java [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.Setter;
+import lombok.ToString;
+
+@Getter
+@Setter
+@ToString
+public class UserResponseDto {
+    private Long id;
+    private String email;
+    private String password;
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/dto/UserResponseDto.java]: Don't add sensitive data to the response body!!!

File: src/main/java/org/example/bookshop/service/UserService.java [hunk 9]
```diff
@@ -0,0 +1,8 @@
+package org.example.bookshop.service;
+
+import org.example.bookshop.dto.UserRegistrationRequestDto;
+import org.example.bookshop.dto.UserResponseDto;
+
+public interface UserService {
+    UserResponseDto save(UserRegistrationRequestDto requestDto);
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/UserService.java]:
```suggestion
    UserResponseDto register(UserRegistrationRequestDto requestDto);
```

File: src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+@RequiredArgsConstructor
+@Service
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto save(UserRegistrationRequestDto requestDto) {
+        if (userRepository.existsByEmail(requestDto.getEmail())) {
+            throw new RegistrationException("User already exist");
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java]: Let's add an email parameter to the message

File: src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        }
+
+        User user = userMapper.toModel(requestDto);
+        User saved = userRepository.save(user);
+        UserResponseDto dto = userMapper.toDto(saved);
+        System.out.println(dto);
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java]: remove printing to the console. If you want to add logs you can use Slf4j

File: src/main/java/org/example/bookshop/controller/AuthenticationController.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+import org.example.bookshop.exception.RegistrationException;
+import org.example.bookshop.service.UserService;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/authorization")
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/controller/AuthenticationController.java]:
```suggestion
@RequestMapping("/auth")
```
Let's follow technical requirements

File: src/main/java/org/example/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 13]
```diff
@@ -11,28 +11,42 @@
...
 import org.springframework.validation.ObjectError;
 import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.ExceptionHandler;
 import org.springframework.web.bind.annotation.RestControllerAdvice;
 import org.springframework.web.context.request.WebRequest;
 import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
 
 @RestControllerAdvice
 public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+    private static final Map<String, Object> body = new LinkedHashMap<>();
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/exception/CustomGlobalExceptionHandler.java]: Don't use the body as a static variable of class. You can have unexpected problems with responses in the future.

File: src/main/java/org/example/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 14]
```diff
@@ -11,28 +11,42 @@
...
+        body.put(ERRORS, errors);
         return new ResponseEntity<>(body, headers, HttpStatus.BAD_REQUEST);
     }
 
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(
+            RegistrationException ex) {
+        body.put(TIMESTAMP, LocalDateTime.now());
+        body.put(STATUS, HttpStatus.CONFLICT);
+        body.put(ERRORS, ex.getMessage());
```
Comment (liliia-ponomarenko) [inline src/main/java/org/example/bookshop/exception/CustomGlobalExceptionHandler.java]: You have a similar code in the previous method, try to move it to the private method ;)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution ;)

File: src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java (hunk 11 above)
Comment (dmarmul) [reply src/main/java/org/example/bookshop/service/impl/UserServiceImpl.java]: my bad, I tried to find mistake and forgot remove it.

File: src/main/java/org/example/bookshop/dto/UserResponseDto.java [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+
+import lombok.Getter;
+import lombok.Setter;
+import lombok.ToString;
+
+@Getter
+@Setter
+@ToString
+public class UserResponseDto {
+    private String email;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/dto/UserResponseDto.java:12]:
```suggestion
    private Long id;
    private String email;
```

File: src/main/java/org/example/bookshop/dto/UserResponseDto.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+    private String firstName;
+    private String lastName;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/dto/UserResponseDto.java:14]:
```suggestion
    private String lastName;
    private String shippingAddress;
```

File: src/main/java/org/example/bookshop/controller/AuthenticationController.java [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+import org.example.bookshop.service.UserService;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/auth")
+public class AuthenticationController {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/bookshop/controller/AuthenticationController.java:20]: Add swagger annotations to controller

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by Elena-Bruyako — APPROVED** [review]
well done

**Pull request merged**